src/mini.lalrpop

```diff
@@ -90,24 +87,85 @@ PathDecl: Vec<String> = {
 }
 
 BodyDecl: TopLevelDecl = {
-	TypeDecl => TopLevelDecl::TypeDecl(<>),
-	FuncDecl => TopLevelDecl::FuncDecl(<>),
-	GlobalVarDecl => TopLevelDecl::VarDecl(<>),
+    TypeDecl => TopLevelDecl::TypeDecl(<>),
+    FuncDecl => TopLevelDecl::FuncDecl(<>),
+    GlobalVarDecl => TopLevelDecl::VarDecl(<>),
 }
 
 TypeDecl: TypeDecl = {
-    "type" <Ident> "=" <Type> ";" => new_type_decl(<>),
+    "type" <i: Ident> <g: Generalization?> "=" <t: Type> ";" =>? {
+        let params: Vec<_> = g.into_iter().flatten().collect();
+        let names = check_generic_parameters(params.clone(), &string_table)?;
+        let tipe = t.make_generic(&names);
+        
+        for (index, (name, debug_info)) in params.into_iter().enumerate() {
+            let unused = tipe.find(&|tipe: &Type| *tipe == Type::GenericSlot(index)).is_empty();
+            if unused {
+                Err(CompileError::new(
+                    "Generics error",
+                    format!(
+                        "Type {}'s generic arg {} isn't used",
+                        Color::red(string_table.name_from_id(i)),
+                        Color::red(string_table.name_from_id(name)),
+                    ),
+                    debug_info.locs(),
+                ))?;
+            }
+        }
+        Ok(new_type_decl(i, tipe))
+    }
 }
 
 FuncDecl: Func = {
-	<attribs: ("#[" <Attributes> "]")?> <qualifiers: (<Qualifier>)*> "func" <lno: @L> <i:Ident> "(" <fa:FuncArgs> ")" <t: (Return)?> <cb:CodeBlock> => {
+    <attribs: ("#[" <Attributes> "]")?> <qualifiers: (<Qualifier>)*> "func" <lno: @L> <i: Ident> <g: Generalization?> "(" <mut args: FuncArgs> ")" <mut ret: Return?> <block: CodeBlock> =>? {
         let public = qualifiers.contains(&"public");
         let view = qualifiers.contains(&"view");
         let write = qualifiers.contains(&"write");
         let debug = DebugInfo::new(file_info.location(BytePos::from(lno), filename), attribs.unwrap_or_default());
-        let name = stringtable.name_from_id(i).clone();
+        let name = string_table.name_from_id(i).clone();
+
+        // Generalize over the generic parameters
+        let params: Vec<_> = g.into_iter().flatten().collect();
+        let generic_ids = check_generic_parameters(params.clone(), &string_table)?;
+        args.iter_mut().for_each(|arg| arg.tipe = arg.tipe.make_generic(&generic_ids));
+        if let Some(ref mut tipe) = ret {
```

Comment:
This is simpler as `Some(tipe) = &mut ret`

Reply:
Sure, we can use that style. I've pushed a change